### ISSUE TITLE
Eval bug: llama-perplexity crashes when generating logits at medium-large context (only when model is split across GPUs/GPU and CPU?)

### ISSUE THREAD

**Issue by compilebunny** [ggml-org/llama.cpp#23490]
### Name and Version

llama.cpp version: 9189 (64b38b561)

### Operating systems

Linux

### GGML backends

CUDA

### Hardware

Nvidia 5060Ti / Intel CPU

### Models

Gemma 4 31B - BF16 GGUF

### Problem description & steps to reproduce

I want to measure perplexity and KL-divergence across a variety of context lengths. The system that performs the measurements has to run the BF16 of Gemma 4 31B split across GPU and RAM. It works fine for context length = 1024 but crashes at context length = 10000. 

command is: 

`
llama-perplexity -m gemma-4-31B-it-BF16-00001-of-00002.gguf -f ./logits_temp/txt_ctx10000_ch0.txt -c 10000 -t 10 --no-mmap --no-warmup -ctk bf16 -ctv bf16 --kl-divergence-base ./logits_temp/gt_gemma-4-31B-it-BF16-00001-of-00002.gguf_ctx10000_ch0.bin --fit off -ngl 13 -dev CUDA0 --verbose`

The main error is:

```
0.47.670.644 I perplexity: saving all logits to ./logits_temp/gt_gemma-4-31B-it-BF16-00001-of-00002.gguf_ctx10000_ch0.bin
0.47.670.648 I perplexity: tokenizing the input ..
0.47.743.853 I perplexity: tokenization took 73.199 ms
0.47.743.922 I perplexity: calculating perplexity over 2 chunks, n_ctx=10000, batch_size=2048, n_seq=1
llama.cpp/build/bin/libggml-base.so.0(+0x1a7b6) [0x7871cf7947b6]
llama.cpp/build/bin/libggml-base.so.0(ggml_print_backtrace+0x20d) [0x7871cf794c3d]
llama.cpp/build/bin/libggml-base.so.0(+0x2e69f) [0x7871cf7a869f]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xc364a) [0x7871ceac364a]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(_ZSt10unexpectedv+0x0) [0x7871ceaabc6c]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xc3901) [0x7871ceac3901]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(_ZSt20__throw_length_errorPKc+0x44) [0x7871ceab0073]
llama-perplexity(+0x4f87) [0x64175a061f87]
llama-perplexity(+0x7f48) [0x64175a064f48]
/usr/lib/x86_64-linux-gnu/libc.so.6(+0x2a601) [0x7871ce62a601]
/usr/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x88) [0x7871ce62a718]
llama-perplexity(+0x82a5) [0x64175a0652a5]
terminate called after throwing an instance of 'std::length_error'
  what():  vector::_M_default_append
Aborted                    (core dumped)
```

### Relevant log output

<details>
<summary>Logs</summary>

```console

```
</details>

**Comment by fairydreaming** [comment]
I confirm the issue, it's caused by integer overflow in llama-perplexity. Here's a simple fix to apply if you can compile from the source:

```
diff --git a/tools/perplexity/perplexity.cpp b/tools/perplexity/perplexity.cpp
index f66576eb4..ea9de83e1 100644
--- a/tools/perplexity/perplexity.cpp
+++ b/tools/perplexity/perplexity.cpp
@@ -524,7 +524,7 @@ static results_perplexity perplexity(llama_context * ctx, const common_params &
         logits_stream.write((const char *)&n_chunk, sizeof(n_chunk));
         logits_stream.write((const char *)tokens.data(), n_chunk*n_ctx*sizeof(tokens[0]));
         const int nv = 2*((n_vocab + 1)/2) + 4;
-        log_probs.resize(n_ctx * nv);
+        log_probs.resize(size_t(n_ctx) * nv);
     }
 
     // We get the logits for all the tokens in the context window (params.n_ctx)
```

**Comment by fairydreaming** [comment]
Full stacktrace for clarity:

```
# ./bin/llama-perplexity -m ~/ggufs/gemma-4-31B-it-f16.gguf -f ../../perplexity/wikitext-2-raw/wiki.test.raw -c 10000 -t 10 --no-mmap --no-warmup -ctk bf16 -ctv bf16 --kl-divergence-base /dev/null --fit off -ngl 99 --verbose
...
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x000073a6fdd10813 in __GI___wait4 (pid=18823, stat_loc=0x0, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
warning: 30	../sysdeps/unix/sysv/linux/wait4.c: No such file or directory
#0  0x000073a6fdd10813 in __GI___wait4 (pid=18823, stat_loc=0x0, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
30	in ../sysdeps/unix/sysv/linux/wait4.c
#1  0x000073a6fdb0e91f in ggml_print_backtrace () at /home/phm/projects/llama.cpp/ggml/src/ggml.c:234
234	        waitpid(child_pid, NULL, 0);
#2  0x000073a6fdb25b6d in ggml_uncaught_exception () at /home/phm/projects/llama.cpp/ggml/src/ggml.cpp:9
9	    ggml_print_backtrace();
#3  0x000073a6fbabb0da in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x000073a6fbaa5a55 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x000073a6fbabb391 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x000073a6fbaa92d2 in std::__throw_length_error(char const*) () from /lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x000073a6fdf6b2f7 in std::vector<unsigned short, std::allocator<unsigned short> >::_M_check_len (this=0x7ffe7f566ad0, __n=18446744072036064320, __s=0x73a6fdf75471 "vector::_M_default_append") at /usr/include/c++/13/bits/stl_vector.h:1899
1899		  __throw_length_error(__N(__s));
#8  0x000073a6fdf667c3 in std::vector<unsigned short, std::allocator<unsigned short> >::_M_default_append (this=0x7ffe7f566ad0, __n=18446744072036064320) at /usr/include/c++/13/bits/vector.tcc:662
662			_M_check_len(__n, "vector::_M_default_append");
#9  0x000073a6fdf61f05 in std::vector<unsigned short, std::allocator<unsigned short> >::resize (this=0x7ffe7f566ad0, __new_size=18446744072036064320) at /usr/include/c++/13/bits/stl_vector.h:1016
1016		  _M_default_append(__new_size - size());
#10 0x000073a6fdf4e540 in perplexity (ctx=0x65191d4be890, params=..., n_ctx=10000) at /home/phm/projects/llama.cpp/tools/perplexity/perplexity.cpp:528
528	        log_probs.resize(n_ctx * nv);
#11 0x000073a6fdf5a860 in llama_perplexity (argc=22, argv=0x7ffe7f5687b8) at /home/phm/projects/llama.cpp/tools/perplexity/perplexity.cpp:2090
2090	        results = perplexity(ctx, params, n_ctx);
#12 0x000065191a19316d in main (argc=22, argv=0x7ffe7f5687b8) at /home/phm/projects/llama.cpp/tools/perplexity/main.cpp:4
4	    return llama_perplexity(argc, argv);
[Inferior 1 (process 18814) detached]
terminate called after throwing an instance of 'std::length_error'
  what():  vector::_M_default_append
Aborted (core dumped)
```
